extensions/amp-a4a/0.1/amp-a4a.js

```diff
@@ -259,52 +259,54 @@ export class AmpA4A extends AMP.BaseElement {
     // promise chain due to cancel from unlayout, the promise will be rejected.
     this.promiseId_++;
     const promiseId = this.promiseId_;
+    // Shorthand for: reject promise if current chain is out of date.
+    const checkStillCurrent = promiseId => {
+      if (promiseId != this.promiseId_) {
+        throw cancellation();
+      }
+    };
     this.adPromise_ = viewerFor(this.win).whenFirstVisible()
+      // The following block returns the ad URL, if one is available.
       .then(() => {
-        if (promiseId != this.promiseId_) {
-          return Promise.reject(cancellation());
-        }
+        checkStillCurrent(promiseId);
         return this.getAdUrl();
       })
+      // This block returns the (possibly empty) response to the XHR request.
       .then(adUrl => {
-        if (promiseId != this.promiseId_) {
-          return Promise.reject(cancellation());
-        }
+        checkStillCurrent(promiseId);
         this.adUrl_ = adUrl;
         return this.sendXhrRequest_(adUrl);
       })
+      // The following block returns either a valid ad creative or null, if
+      // no valid creative is available.
       .then(fetchResponse => {
-        if (promiseId != this.promiseId_) {
-          return Promise.reject(cancellation());
+        checkStillCurrent(promiseId);
+        if (!fetchResponse || !fetchResponse.arrayBuffer) {
+          return Promise.resolve(null);
         }
-        if (fetchResponse && fetchResponse.arrayBuffer) {
-          return fetchResponse.arrayBuffer().then(bytes => {
-            if (promiseId != this.promiseId_) {
-              return Promise.reject(cancellation());
-            }
-            return this.extractCreativeAndSignature(
-                bytes, fetchResponse.headers).then(responseParts => {
-                  if (promiseId != this.promiseId_) {
-                    return Promise.reject(cancellation());
-                  }
-                  return this.validateAdResponse_(
-                      responseParts.creative, responseParts.signature)
-                      .then(valid => {
-                        if (promiseId != this.promiseId_) {
-                          return Promise.reject(cancellation());
-                        }
-                        // Note: It's critical that #maybeRenderAmpAd_ be called
-                        // on precisely the same creative that was validated
-                        // via #validateAdResponse_.  See GitHub issue
-                        // https://github.com/ampproject/amphtml/issues/4187
-                        return this.maybeRenderAmpAd_(valid,
-                            responseParts.creative);
-                      });
-                });
-          });
-        } else {
+        return fetchResponse.arrayBuffer().then(bytes => {
+          checkStillCurrent(promiseId);
+          return this.extractCreativeAndSignature(bytes,
+              fetchResponse.headers);
+        })
+        .then(responseParts => {
```

Comment:
This is doubly nested (`#then` inside a `#then`) when it doesn't need to be. Please bring it up to the top level.